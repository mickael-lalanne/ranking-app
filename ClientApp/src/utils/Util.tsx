import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
