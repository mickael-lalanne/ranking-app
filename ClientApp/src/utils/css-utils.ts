export const ELEMENT_SIZE: string = '75px';
