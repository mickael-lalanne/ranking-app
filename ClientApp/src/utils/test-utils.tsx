import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
