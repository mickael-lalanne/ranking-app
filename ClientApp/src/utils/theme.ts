import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
