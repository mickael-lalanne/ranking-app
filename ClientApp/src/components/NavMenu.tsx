import React, { useState } from 'react';
import {
    Collapse,
    Navbar,
    NavbarBrand,
    NavbarToggler,
    NavItem,
    NavLink,
} from 'reactstrap';
import { Link } from 'react-router-dom';
import './NavMenu.css';
import { css } from '@emotion/css';
import LinearProgress from '@mui/material/LinearProgress';
import { useAppSelector } from '../app/hooks';
import { THEMES_COLORS } from '../utils/theme';
import { SignOutButton } from '@clerk/clerk-react';

const NavMenu = (
    { height }: { height: string}
) => {
    const [collapsed, setCollapsed] = useState<boolean>(true);

    const loading: boolean = useAppSelector(state => state.application.loading);

    const toggleNavbar = () =>  {
        setCollapsed(!collapsed);
    }

    const ShowLoading = (): React.JSX.Element | undefined => {
        if (loading) {
            return <LinearProgress
                color="primary"
                className={loading_indicator_style}
                style={{ position: 'absolute' }}
            />;
        }
    };

    return (
        <header style={{ minHeight: height }} className={ loading ? nav_disabled_style : '' }>
            <Navbar
                className="navbar-expand-sm navbar-toggleable-sm ng-white border-bottom box-shadow mb-3"
                container
                light
            >
                <NavbarBrand tag={Link} to="/">
                    Ranking App
                </NavbarBrand>
                <NavbarToggler
                    onClick={toggleNavbar}
                    className="mr-2"
                />
                <Collapse
                    className="d-sm-inline-flex flex-sm-row-reverse"
                    isOpen={!collapsed}
                    navbar
                >
                    <ul className="navbar-nav flex-grow">
                        <NavItem>
                            <NavLink
                                tag={Link}
                                className={navlink_style}
                                to="/"
                            >
                                Home
                            </NavLink>
                        </NavItem>
                        <NavItem>
                            <NavLink
                                tag={Link}
                                className={navlink_style}
                                to="/templates"
                            >
                                Templates
                            </NavLink>
                        </NavItem>
                        <NavItem>
                            <NavLink
                                tag={Link}
                                className={navlink_style}
                                to="/tierlists"
                            >
                                Tierlists
                            </NavLink>
                        </NavItem>
                        <NavItem>
                            <NavLink className={navlink_style}>
                                <SignOutButton>
                                    <span>Logout</span>
                                </SignOutButton>
                            </NavLink>
                        </NavItem>
                    </ul>
                </Collapse>
            {ShowLoading()}
            </Navbar>
        </header>
    );
}
export default NavMenu;

/**
 * CSS STYLES
 */
const navlink_style = css({
    color: 'black',
    cursor: 'pointer',
    '&:hover': {
        color: THEMES_COLORS?.primary,
        textDecoration: 'underline'
    }
});

const loading_indicator_style = css({
    bottom: '0',
    left: 0,
    width: '100%'
});

const nav_disabled_style = css({
    pointerEvents: 'none',
    userSelect: 'none',
    opacity: 0.5
});
