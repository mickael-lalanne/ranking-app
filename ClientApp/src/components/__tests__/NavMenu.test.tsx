import * as React from 'react';
import NavMenu from '../NavMenu';
import { renderWithProviders } from '../../utils/test-utils';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/react';

const clerkPath: string = '@clerk/clerk-react';
const mockSignOutButton = jest.fn(children => children);
jest.mock(clerkPath, () => ({
    ...jest.requireActual(clerkPath),
    SignOutButton: ({ children }: React.PropsWithChildren<unknown>) => {
        return mockSignOutButton(children);
    },
}));

const NavMenuTest = () => {
    return (
        <MemoryRouter>
            <NavMenu height='60' />
        </MemoryRouter>
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(NavMenuTest());

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    it('clicks on logout button', async () => {
        renderWithProviders(NavMenuTest());

        await userEvent.click(screen.getByText('Logout'));
        expect(mockSignOutButton).toHaveBeenCalled();
    });
});

// FYI: Routers tests are made in App.tsx but also in e2e tests with cypress
