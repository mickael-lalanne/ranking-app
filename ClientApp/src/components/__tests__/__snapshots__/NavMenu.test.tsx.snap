// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <header
      class=""
      style="min-height: 60;"
    >
      <nav
        class="navbar-expand-sm navbar-toggleable-sm ng-white border-bottom box-shadow mb-3 navbar navbar-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            <img
              class="css-1qu68v9"
              src="[object Object]"
            />
            Ranking App
          </a>
          <button
            aria-label="Toggle navigation"
            class="mr-2 navbar-toggler"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="d-sm-inline-flex flex-sm-row-reverse collapse navbar-collapse"
          >
            <ul
              class="navbar-nav flex-grow"
            >
              <li
                class="nav-item"
              >
                <a
                  class="css-8dyutd nav-link"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="css-8dyutd nav-link"
                  href="/templates"
                >
                  Templates
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="css-8dyutd nav-link"
                  href="/tierlists"
                >
                  Tierlists
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="css-8dyutd nav-link"
                >
                  <span>
                    Logout
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </header>
  </div>
</body>
`;
